clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -287,12 +291,16 @@ public RecordAppendResult append(String topic,
                     }
                 }
 
-                // we don't have an in-progress record batch try to allocate a new batch
-                if (abortOnNewBatch) {
-                    // Return a result that will cause another call to append.
+                // noDqForPartition is true either when 1. partition was encountered for first time so no Deque existed previously.
+                // 2. DQ was removed due to - all batches were cleared due to expiration or sender cleared batches after draining.
+                // if so, abort and look to call partitioner -> onNewBatch and select other partition.
+                // This prevents a single partition getting re-selected after recent drain.
+                if (abortOnNewBatch && noDqForPartition) {
```

Comment:
I am not sure if this completely addresses the issue. For example, if the producer adds records relatively slowly, it's possible that by the time the RoundRobinPartitioner adds the record to the same partition again, the batches for that partition have already been drained.

Here is another possibility. We could change the computation for the following code in KafkaProducer. If a partitioner is defined, but doesn't override the default implementation of `onNewBatch`(which is the case for RoundRobinPartitioner), we could  also set abortOnNewBatch to false. We could use reflection to get the method and do the `isDefault` test.

```
            // A custom partitioner may take advantage on the onNewBatch callback.
            boolean abortOnNewBatch = partitioner != null;

```

What do you think @jolshan ?

Reply:
> if a partitioner is defined, but doesn't override the default implementation of onNewBatch

I think this is the only correct way to fix this without breaking partitioners that depend on the onNewBatch being called to switch partitions.  Given that the onNewBatch is deprecated now, we should avoid abortOnNewBatch logic unless it's defined.

Reply:
Is this an issue for all partitioners or just `RoundRobinPartitioner`?

Reply:
This is potentially an issue for any partitioner not overriding the default empty implementation for `onNewBatch` since the partitioner is unnecessarily called a second time for the same record on appending to a new batch.

Reply:
The old sticky partitioner changed the behavior such that `partition` is called one more time when a batch is created (which is not a problem for the old sticky partitioner, because `partition` isn't changing state, the state is changed by `onNewBatch`) but changes behavior of pretty much all other partitioners that change state on `partition`.  The `RoundRobinPartitioner` probably has the most visible effects, but all partitioners that change state on `partition` are affected.

It's kind of surprising that this change of behavior wasn't noticed for a while (which probably indicates how often custom partitioners are used).

Eliminating the second call of `partition` would change the partition switching pattern for the old sticky partitioner, and we didn't want to do that in case people want to switch back to the default logic.  So doing the reflection lookup is probably the best way to separate when we need the `onNewBatch` & second `partition` call, vs. doing one `partition` call per `tryAppend`.

Reply:
It actually was noticed back in 2020. :( I just missed the comments. https://github.com/apache/kafka/commit/717c55be971df862c55f55d245b9997f1d6f998c#r39200381 
We also saw a ticket filed for this yesterday. https://issues.apache.org/jira/browse/KAFKA-14573

Reply:
Using reflection is a creative option, but it does mean that if someone overrides the method to eg add a log, then suddenly things will break. Which would be quite surprising (it goes without saying).

Reply:
Compatibility is tricky here. I think back when we implemented this I remember there were some concerns about the onNewBatch method since we wanted to avoid too many changes to the interface and allow for old partitioners to work as intended. I suppose if there was a way to avoid "abortOnNewBatch" and the subsequent call to partition, that would be ideal, but not sure it is something we can add now.

Reply:
`onNewBatch` is now deprecated, so hopefully, people would take a hint not to override it :-)

But, yeah, agree with calling it creative (I wouldn't have come up with that given my C++ background -- these ideas are blocked before they appear in my mind).